joblib/shelf.py

```diff
@@ -0,0 +1,237 @@
+
+        Returns:
+        --------
+            A future on the shelved data.
+
+        """
+        data_id = uuid4().hex
+        self.store_backend.dump_item([data_id], data)
+
+        future = JoblibShelfFuture(data_id, self)
+
+        def _collect(ref):
+            self._on_collect(data_id)
+        self._futures.add(weakref.ref(future, _collect))
+        return future
+
+    def _on_collect(self, data_id):
+        self.store_backend.clear_location(
+            os.path.join(self.store_backend.location, data_id))
+
+    def __del__(self):
+        # TODO: FIX self.clear()
+        shutil.rmtree(self.store_backend.location)
+
+
+def shelve(input_object):
+    """Shelves an arbitrary object and returns a future on it.
+
+    The input object can then be deleted at any time by the script to save
+    memory. The future, a light-weight object, can be used later to reload the
+    initial object.
+
+    During the life of the future, the input object is kept written on a store
+    (by default a file on a disk).
+    To retrieve the original input object later, use the ``result`` method of
+    the returned future, this call will reload the initial data from the disk
+    and return it.
+
+    Parameters
+    ----------
+    input_object: any
+        The input object to shelve.
+
+    Returns
+    -------
+        A future referencing the shelved object
+
+    See Also
+    --------
+        joblib.shelve_mmap : function to shelve a numpy array. The future
+        result returns a memmapped numpy array.
+
+    Notes
+    -----
+        The content of the shelved object is effectively deleted from the store
+        only when no reference on its future no longer exists.
+        When the interpreter process exits, all shelved objects are deleted
+        from the store.
+
+    Examples
+    --------
+
+        A simple example:
+
+        >>> from joblib import shelve
+        >>> input_object = "The object on the shelf"
+        >>> future = shelve(input_object)
+        >>> del input_object  # input object can now be removed from memory
+        >>> future  #doctest: +SKIP
+        <joblib.shelf.JoblibShelfFuture at 0x7fc15fad7a90>
+        >>> print(future.result())
+        The object on the shelf
+
+        Only the 'threading' backend of `Parallel` can be used with `shelve`:
+
+        >>> from joblib import shelve, Parallel, delayed
+        >>> def func(future):
+        ...     return future.result()**2
+        >>> r = Parallel(n_jobs=10, backend='threading')(
+        ...              delayed(func)(shelve(i)) for i in range(10))
+        >>> r
+        [0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
+
+    """
+    global _active_shelf
+    if _active_shelf is None:
+        tmp_folder = os.environ.get('JOBLIB_TEMP_FOLDER',
+                                    tempfile.gettempdir())
+        _active_shelf = JoblibShelf(tmp_folder)
+
+    return _active_shelf.put(input_object)
+
+
+def shelve_mmap(input_array):
```

Comment:
I think that the variable should rather be called "input_object" (here and in the docstring below, the word "array" should often be replaced by "object").

Reply:
This function is only meant to used with numpy array, since it return a future on a mmap. That's why I think it's important to keep the 'array' in the variable name